### PR TITLE
chore: improve handling with missing xserver

### PR DESCRIPTION
**Pull request by mxschmitt**
We could improve it in different ways:
- a) if `headless: false` and no `process.env.DISPLAY` configured and os is linux , then yell
- b) if `headless: false` and no `process.env.DISPLAY` configured and os is linux , then automatically try to start a xserver
- c) if browser does not start because of missing xserver (we compare the error message), then yell 

Currently c) is implemented in this PR, but I'm probably more for a). Not sure how reliable this is tho.

Relates: 
- https://github.com/microsoft/playwright/issues/9888
- https://github.com/microsoft/playwright/discussions/12704
- https://github.com/microsoft/playwright/issues/6037

### CODE REVIEW

File: packages/playwright-core/src/server/browserType.ts [hunk 1]
```diff
@@ -76,14 +77,28 @@ export abstract class BrowserType extends SdkObject {
...
+        // Chromium
+        errorMessage.includes('Missing X server') ||
+        // WebKit
+        errorMessage.includes('cannot open display') ||
+        // Firefox
+        errorMessage.includes('no DISPLAY environment variable specified')
+      ) {
+        rewriteErrorMessage(error, '\n' + wrapInASCIIBox(
+            'Looks like you launched a headed browser without having a XServer running.\n' +
+          'Set either \'headless: false\' or use \'xvfb-run <your-playwright-app>\' before running Playwright'
```
Comment (rwoll) [inline packages/playwright-core/src/server/browserType.ts:98]: Is it possible to get here on Windows? If so the Xvfb tip might be misleading.

**Review by rwoll — COMMENTED** [review]
Overall looks good and nice to have a more explicit/friendly error for users to debug.

Option (b) of auto-adding Xvfb when needed would be particularly helpful on GitHub Codespaces.

**Comment by yury-s** [comment]
> (a) and (b) rely on DISLPAY env variable. Do all window servers (e.g. wayland) expose it?

Looks like at least sometimes it can be [WAYLAND_DISPLAY](https://wayland-book.com/wayland-display/creation.html).

File: packages/playwright-core/src/server/browserType.ts (hunk 1 above)
Comment (mxschmitt) [reply packages/playwright-core/src/server/browserType.ts:98]: there is no such a thing in Windows, so I doubt it.

**Comment by aslushnikov** [comment]
If https://github.com/microsoft/playwright/pull/12806 lands, we can auto-run xvfb inside of our docker container.

File: packages/playwright-core/src/server/browserType.ts [hunk 2]
```diff
@@ -76,14 +77,28 @@ export abstract class BrowserType extends SdkObject {
...
       // @see https://github.com/microsoft/playwright/issues/5214
-      const errorMessage = typeof error === 'object' && typeof error.message === 'string' ? error.message : '';
+      const errorMessage: string = typeof error === 'object' && typeof error.message === 'string' ? error.message : '';
       if (errorMessage.includes('Inconsistency detected by ld.so')) {
         progress.log(`<restarting browser due to hitting race condition in glibc>`);
         return this._innerLaunch(progress, options, persistent, protocolLogger, userDataDir);
       }
+      if (
+        // Chromium
+        errorMessage.includes('Missing X server') ||
```
Comment (aslushnikov) [inline packages/playwright-core/src/server/browserType.ts:90]: Shouldn't this all go to `_rewriteStartupError`?

Comment (mxschmitt) [reply packages/playwright-core/src/server/browserType.ts]: done

**Comment by mxschmitt** [comment]
Error looks now like this:

```
root@abc4a32a047f:/work# node test.js
node:internal/process/promises:265
            triggerUncaughtException(err, true /* fromPromise */);
            ^

browserType.launch: 
╔════════════════════════════════════════════════════════════════════════════════════════════════╗
║ Looks like you launched a headed browser without having a XServer running.                     ║
║ Set either 'headless: false' or use 'xvfb-run <your-playwright-app>' before running Playwright ║
║                                                                                                ║
║ <3 Playwright Team                                                                             ║
╚════════════════════════════════════════════════════════════════════════════════════════════════╝
=========================== logs ===========================
<launching> /ms-playwright/chromium-980066/chrome-linux/chrome --disable-background-networking --enable-features=NetworkService,NetworkServiceInProcess --disable-background-timer-throttling --disable-backgrounding-occluded-windows --disable-breakpad --disable-client-side-phishing-detection --disable-component-extensions-with-background-pages --disable-default-apps --disable-dev-shm-usage --disable-extensions --disable-features=ImprovedCookieControls,LazyFrameLoading,GlobalMediaControls,DestroyProfileOnBrowserClose,MediaRouter,AcceptCHFrame,AutoExpandDetailsElement --allow-pre-commit-input --disable-hang-monitor --disable-ipc-flooding-protection --disable-popup-blocking --disable-prompt-on-repost --disable-renderer-backgrounding --disable-sync --force-color-profile=srgb --metrics-recording-only --no-first-run --enable-automation --password-store=basic --use-mock-keychain --no-service-autorun --export-tagged-pdf --no-sandbox --user-data-dir=/tmp/playwright_chromiumdev_profile-nzgcMS --remote-debugging-pipe --no-startup-window
<launched> pid=306
[pid=306][err] [306:306:0318/180622.264244:ERROR:ozone_platform_x11.cc(247)] Missing X server or $DISPLAY
[pid=306][err] [306:306:0318/180622.264665:ERROR:env.cc(225)] The platform failed to initialize.  Exiting.
============================================================
    at /work/test.js:4:34
    at Object.<anonymous> (/work/test.js:7:3) {
  name: 'Error'
}
```

File: packages/playwright-core/src/server/browserType.ts [hunk 3]
```diff
@@ -76,7 +79,7 @@ export abstract class BrowserType extends SdkObject {
 
   async _innerLaunchWithRetries(progress: Progress, options: types.LaunchOptions, persistent: types.BrowserContextOptions | undefined, protocolLogger: types.ProtocolLogger, userDataDir?: string): Promise<Browser> {
     try {
-      return this._innerLaunch(progress, options, persistent, protocolLogger, userDataDir);
+      return await this._innerLaunch(progress, options, persistent, protocolLogger, userDataDir);
```
Comment (mxschmitt) [inline packages/playwright-core/src/server/browserType.ts:82]: I have the assumption that the actual stuff underneath with checking the error was never working. This should fix it!

**Review by aslushnikov — APPROVED** [review]
awesome! Can we have a test?

**Pull request merged**